(#357) Add Export Command

gep13 opened the pull request:
This will allow the creation of a packages.config file of all the
currently installed packages on the machine.  Usage of this command
will be:

choco export -o c:/temp/packages.config --include-version-numbers

This command is particularly useful when re-building a machine.  i.e.
First export all packages currently installed on machine, and then
replay this packages.config via choco install packages.config on new
machine.

Fixes #357

RichiCoder1 commented:
Does this/could this include pinned packages info? And/or installation/package args?

gep13 commented:
There is no reason that additional features couldn’t be added to this new command. I would see that as a separate enhancement issue personally though, as there would be design decisions that would need to be made around the persisted arguments. @ferventcoder and I have spoken briefly about changes that need to happen in this area, and things would need to be flushed out.

Regarding the pinned package side of things, are you asking that if a pinned is pinned, this makes it into the packages.config file and when replayed on a new machine, that package also ends up being pinned?

gep13 commented:
@RichiCoder1

RichiCoder1 commented:
> There is no reason that additional features couldn’t be added to this new command. I would see that as a separate enhancement issue personally though, as there would be design decisions that would need to be made around the persisted arguments. @ferventcoder and I have spoken briefly about changes that need to happen in this area, and things would need to be flushed out.

Fair, that makes sense.

> Regarding the pinned package side of things, are you asking that if a pinned is pinned, this makes it into the packages.config file and when replayed on a new machine, that package also ends up being pinned?

Exactly. So if I pinned VS Code on one machine, it would end up pinned on another machine.

gep13 commented:
> @RichiCoder1 said...
Exactly. So if I pinned VS Code on one machine, it would end up pinned on another machine

In the current implementation, pin is a standalone command, I.e. you can’t pin a package during the process of installation. As such, this may need to be implemented as a second pass of the packages.config file to look for pinned packages, or a modification to the install command to also pin when installing.

Again, more discussion would need to be had on this topic.

ferventcoder commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+        public void configure_argument_parser(OptionSet optionSet, ChocolateyConfiguration configuration)
+        {
+            optionSet
+                .Add("o=|output-file-path=",
```
Do we use `o` anywhere else with regards to output?

Additionally we use `output` in a few places, let's keep this consistent.

devlead replied:
![](https://media3.giphy.com/media/Xw6yFn7frR3Y4/giphy.gif)

gep13 replied:
@ferventcoder no, this is the first place that `o` is introduced as a short option, so I think we are good to use it, if you agree.

vexx32 replied:
Does this need to be an option switch? To me reading it, it feels kind of like i'd expect `choco export ./packages.config` to work without requiring an `-o`. The filepath here feels like it should be the primary argument?

gep13 replied:
That is a good point. In order to be in-keeping with the rest of the Chocolatey commands, I have added the ability to either provide the explicit option, or not.  So the following all work:

```
    choco export
    choco export "'c:\temp\packages.config'"
    choco export -o="'c:\temp\packages.config'"
    choco export --output-file-path="'c:\temp\packages.config'"
```

ferventcoder commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+                .Add("o=|output-file-path=",
+                     "Output File Path - the path to where the list of currently installed packages should be saved. Defaults to packages.config.",
+                     option => configuration.ExportCommand.OutputFilePath = option.remove_surrounding_quotes())
+                .Add("include-version-numbers",
```
Could we reuse `--version` here?

gep13 replied:
No :smile:

ferventcoder replied:
Right on, makes sense. I wasn't thinking reusing `version` really explained what this was doing.

AdmiringWorm replied:
Using `--include-version` would maybe make sense?
Less verbose, but the meaning isn't lost.

ferventcoder commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+
+        public void handle_additional_argument_parsing(IList<string> unparsedArguments, ChocolateyConfiguration configuration)
+        {
+            if (unparsedArguments.Count > 0)
```
There is a feature switch to ignore unparsed arguments, we should ensure that is on before throwing an error.

gep13 replied:
Having thought about this, throwing an error here doesn't make any sense.  Original code borrowed from pin command, which was handling parsing of subcommand, which is required here, so let's remove this code.

mwallner commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+
+This is especially helpful when re-building a machine that was created
+using Chocolatey.  Export all packages to a file, and then re-install
+those packages onto new machine using `choco install packages.config`.
```
typo? - https://github.com/gep13/choco/pull/1

gep13 replied:
@mwallner agreed.  Minor tweak required on your PR, and then I can get it pulled in.

mwallner commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+ - 0: operation was successful, no issues detected
+ - -1 or 1: an error has occurred
+
+If you find other exit codes that we have not yet documented, please
```
typo? - https://github.com/gep13/choco/pull/1

gep13 replied:
@mwallner actually, no.  This part does read ok, can you update your PR to not include this change.

ferventcoder commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+
+If you find other exit codes that we have not yet documented, please
+ file a ticket so we can document it at
+ https://github.com/chocolatey/choco/issues/new/choose.
```
Very nice! I see you've already brought in the new exit codes!

ferventcoder commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+            "chocolatey".Log().Info(@"
+    choco export
+    choco export --include-version-numbers
+    choco export --output-file-path=""c:\temp\packages.config""
```
Here we should use best practices, so `--output-file-path=""'c:\temp\packages.config'""` (adding the apostrophes).

gep13 replied:
Agreed, will implement this.

ferventcoder commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,185 @@
+        public void run(ChocolateyConfiguration configuration)
+        {
+            var packageResults = _nugetService.get_all_installed_packages(configuration);
+            var settings = new XmlWriterSettings { Indent = true, Encoding = new UTF8Encoding(false) };
```
Would like to be able to include other items to build out a bigger packages.config (for future use). One thing I would like to see here at the very least is to refactor to IXmlService and XmlService.

gep13 replied:
@ferventcoder do you remember what the thought process was here again?  Looking at the code again just now, and I am not immediately clear on why this would help here.  Thanks

gep13 replied:
I have went through some notes that I had on this, as well as watching back the videos of the Twitch streams that we did on this, and the issue here was (I believe) that if we go down this route, there was no direct way of including/excluding specific attributes within the resulting XML file.  Not without implementing a custom XML Serializer/DeSerializer for the class.  I would suggest that this might be something that we could look at if/when we bring in the ability to include additional attributes into the resulting XML file, but for now, I think the code that we have here is a good compromise.

ferventcoder commented on src/chocolatey/infrastructure.app/configuration/ChocolateyConfiguration.cs:
```diff
@@ -544,4 +547,17 @@ public sealed class ProxyConfiguration
         public bool BypassOnLocal { get; set; }
     }
+
+    [Serializable]
```
Nice.

ferventcoder commented on src/chocolatey/infrastructure.app/configuration/ChocolateyConfiguration.cs:
```diff
@@ -544,4 +547,17 @@ public sealed class ProxyConfiguration
+    {
+        public ExportCommandConfiguration()
+        {
+            OutputFilePath = "packages.config";
```
Do we do this other places?

gep13 replied:
Yip, in at least one other place: https://github.com/chocolatey/choco/blob/master/src/chocolatey/infrastructure.app/configuration/ChocolateyConfiguration.cs#L410

gep13 replied:
Since this is done elsewhere, I think we are good to do this here as well.

mwallner commented:
is there a timeline for this feature to become available? :-)

gep13 commented:
@mwallner I have some tidy up work that I need to do on this PR before it would be ready to be merged in.  I haven't forgotten about it, just need to get round to it.  So, to answer your question, no, there isn't a timeline 😄

gep13 commented:
@ferventcoder I believe this is ready for another review.  There is one outstanding question I believe with regard to the use of XmlService, which I am not clear on.  If this is required, can you let me know exactly how you see this working?  Thanks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/chocolatey/choco?pullRequest=1825) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/chocolatey/choco?pullRequest=1825) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/chocolatey/choco?pullRequest=1825) it.</sub>

eabase commented:
@ferventcoder 
Maybe it's time to merge this?

vexx32 reviewed (approved):
Overall this looks good to me, bar one question I had around the design here: https://github.com/chocolatey/choco/pull/1825#discussion_r616830796

gep13 commented:
@vexx32 I have updated this PR based on your feedback.  If you could take another look, I would appreciate it.  I have also added some Unit Tests since the last review.  Thanks

vexx32 commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,197 @@
+﻿// Copyright © 2017 - 2018 Chocolatey Software, Inc
```
I guess not specifically related to this PR, but should we look at updating the license / copyright dates for this repo at some point? 🤔

Should this have a newer date since it's a newer file? Not sure what the process is for these tbqh.

gep13 replied:
I actually had the same thought about this as I was adding it in....

I have created an issue to track getting this done ahead of the next release: https://github.com/chocolatey/choco/issues/2261 @ferventcoder has a method for bulk updating these, so will get with him about how that is done.

vexx32 commented on src/chocolatey/infrastructure.app/commands/ChocolateyExportCommand.cs:
```diff
@@ -0,0 +1,197 @@
+                configuration.ExportCommand.OutputFilePath = unparsedArguments[0];
+            }
+
+            // If no value has been provided for the OutputFilePath, default to packages.config
+            if (string.IsNullOrWhiteSpace(configuration.ExportCommand.OutputFilePath))
+            {
+                configuration.ExportCommand.OutputFilePath = "packages.config";
+            }
```
If I call `choco export` with no other parameters, will this code be reached? Does `handle_additional_argument_parsing()` get called if there are no additional arguments to parse?

gep13 replied:
Yes, this method is always executed.

vexx32 commented:
Apart from those two things, everything here looks good to me!

vexx32 reviewed (approved):
Looks good to me, nice work! 💖
